### PR TITLE
Mobile client and mobile service integrations

### PR DESCRIPTION
**Pull request by sedroche**
Ticket: https://issues.jboss.org/browse/AEROGEAR-2422

This PR is adding the integration component to the Mobile Client component.
client-integrations

I have reused the service integration component from this [PR](https://github.com/openshift/origin-web-console/pull/2897).

- Extract existing service Integration component into reusable "mobile-association" component
    * This new component will handle provisioning/de-provisioning services and creating/deleting binding and associated states.
- New MobileClientIntegrations component using mobile-association
- Refactor MobileServiceIntegrations component to use mobile-association

# Verification
Provisioning/De-provision/ and create/delete integration states should be the same as outlined in the [PR](https://github.com/openshift/origin-web-console/pull/2897).

## NOTE
Pull down this [branch](https://github.com/sedroche/origin-web-common/tree/add-meta-params-to-binding) of the origin web common repo. Follow these [steps](https://github.com/sedroche/origin-web-console#contributing-to-the-primary-repositories) to link your local origin-web-common with the web-console. 

#### Create
- Client name should be prepopulated on the bind screen
- A mobile client integration should create a secret that contains the service information.
#### Delete
- The secret should be deleted
#### Create
- Service name should be prepopulated on the bind screen
- A mobile service integration should create a secret that contains the service information.
- A podpreset should be created (Resources -> Other resources -> filter for podpreset)
- When the binding is ready the service should be redeployed
- The secret should be mounted on the target service pod under `/etc/secrets/<service-name>` .
#### Delete
- The secret should be deleted.
- The podpreset should be deleted.
- The service should be redeployed
- The secret should be unmounted from the pod

### CODE REVIEW

File: app/scripts/directives/mobileClientIntegration.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  function MobileClientIntegration() {
+    var ctrl = this;
+
+    ctrl.$onInit = function() {
+      ctrl.consumerInstance = ctrl.mobileClient;
+      ctrl.parameterData = {
+        CLIENT_NAME: _.get(ctrl.mobileClient, 'metadata.name')
+      };
+      ctrl.projectName = _.get(ctrl.mobileClient, 'metadata.namespace');
```
Comment (JameelB) [inline app/scripts/directives/mobileClientIntegration.js]: Where's this `projectName` used for?

File: app/scripts/directives/mobileIntegration.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    DataService
+  ) {
+    var ctrl = this;
+    var instancePreferredVersion = APIService.getPreferredVersion('serviceinstances');
+    var bindingPreferredVersion = APIService.getPreferredVersion('servicebindings');
+    var isBindingReady = $filter('isBindingReady');
+    var watches = [];
+
+    ctrl.$onInit = function() {
+      var context = {namespace: ctrl.projectName = _.get(ctrl, 'consumerInstance.metadata.namespace')};
```
Comment (JameelB) [inline app/scripts/directives/mobileIntegration.js]: What is this `projectName` used for?

**Comment by JameelB** [comment]
**Client Integration Verification:**
 Verified for Android, iOS and Cordova
- [x] Able to see a list of possible integrations within the mobile client view when a mobile client is provisioned.
- [x] The correct client name is pre-populated on the bind screen.
- [x] Creates a service binding successfully. 
- [x] Creates a secret for the service binding which contains the correct service information.
- [ ] A pending message is displayed when the integration is being created.
- [ ] The correct information of the integration is shown on the integration section of the mobile client.
- [ ] An option to delete the integration is shown within the integration section of the mobile client.
- [x] Service binding can be successfully deleted. (bindings and secrets created are deleted)

Issues: 
- Does not show the correct information within the mobile client integrations section when a binding is created/being created. (Does not show status pending/integration details when it's completed)
- Cannot delete integration from the mobile client integration as wrong information is shown within the integrations section of the client view 

**Service Integration Verification:**
- [x] A list of possible integrations are shown in the integrations section within the service view.
- [x] The service name is correctly pre-populated on the bind screen.
- [ ] A pending message is displayed when the integration is being created.
- [ ] The correct information of the integration is shown on the integration section of the mobile service.
- [ ] An option to delete the integration is shown within the integration section of the mobile service.
- [x] Successfully creates a service binding and a pod preset once the integration has completed.
- [x] The secret created for the service binding has the correct service information.
- [ ] The service is re-deployed and the secret is mounted on the target service pod when the binding is ready.
- [x] Successfully deleted the service binding and the pod preset as well as the secret created.
- [ ] Service is successfully redeployed on deletion of the integration and the secret is unmounted from the target service pod.

Issues: 
- Does not show the correct information within the mobile integrations within the service overview when an integration is created/is being created.
- Cannot delete integration from the mobile integrations as the wrong information is shown within the integrations section of the service view.

File: app/scripts/directives/mobileClientIntegration.js (hunk 1 above)
Comment (sedroche) [reply app/scripts/directives/mobileClientIntegration.js]: Removed thanks

File: app/scripts/directives/mobileIntegration.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      if (changes.consumerInstance && changes.consumerInstance.currentValue &&
+          changes.integrationServiceClass && changes.integrationServiceClass.currentValue &&
+          !ctrl.watchesSet) {
+        var context = {namespace: _.get(ctrl, 'consumerInstance.metadata.namespace')};
+
+        DataService.list(instancePreferredVersion, context, function(serviceInstancesData) {
+          var data = serviceInstancesData.by('metadata.name');
+          ctrl.checkIntegrationInstance(data);
+
+          watches.push(DataService.watch(bindingPreferredVersion, context, function(bindingData) {
```
Comment (maleck13) [inline app/scripts/directives/mobileIntegration.js]: As this happen onchange are we potentially adding more and more watches here? Not sure if we are just want to ask the question.

File: app/scripts/directives/mobileIntegration.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    ctrl.openIntegrationPanel = function() {
+      ctrl.integrationPanelVisible = true;
+    };
+
+    ctrl.provision = function() {
+      $scope.$emit('open-overlay-panel', Catalog.getServiceItem(ctrl.integrationServiceClass));
+    };
+
+    ctrl.$onDestroy = function() {
```
Comment (maleck13) [inline app/scripts/directives/mobileIntegration.js]: ahh i see we are removing them here, but still wondering about the pattern above where we add a watch on each change?

File: app/scripts/directives/mobileIntegration.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      if (!ctrl.binding) {
+        ctrl.hasBinding = false;
+        ctrl.isBindPending = false;
+      }
+      if (ctrl.binding && ctrl.binding.status.currentOperation === 'Unbind') {
+        ctrl.hasBinding = false;
+        ctrl.isBindPending = true;
+      }
+    };
+
```
[the remark this answers is not in the document]
Comment (sedroche) [reply app/scripts/directives/mobileIntegration.js]: :+1: Changed to `.toLowerCase`

File: app/scripts/directives/mobileIntegration.js (hunk 3 above)
Comment (sedroche) [reply app/scripts/directives/mobileIntegration.js]: The `ctrl.watchesSet` flag is used to only set the watches once.

**Pull request closed without merging**